File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.py
def solution(record):
    answer = []
    
    name_dict = {}
    
    for i in record:
        if not i.startswith('Leave'):
            state, uid, name = i.split()
        name_dict[uid] = name

    for i in record:
        if i.startswith('Enter'):
            state, uid, name = i.split()
            answer.append(name_dict[uid]+"님이 들어왔습니다.")
        elif i.startswith('Leave'):
            state, uid = i.split()
            answer.append(name_dict[uid]+"님이 나갔습니다.")
    
    return answer

File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.java
import java.util.*;

class Solution {
    public String[] solution(String[] record) {
        Map<String, String> user = new HashMap<>();
        List<String> action = new ArrayList<>();
        List<String> userId = new ArrayList<>();
        
        for (String r : record) {
            String[] parts = r.split(" ");
            String act = parts[0];
            String uid = parts[1];
            if (!act.equals("Change")) {
                action.add(act);
                userId.add(uid);
            }
            if (parts.length == 3) {
                String name = parts[2];
                user.put(uid, name);
            }
        }
        
        int size = action.size();
        String[] answer = new String[size];
        for (int i = 0; i < size; i++) {
            String userName = user.get(userId.get(i));
            String result = userName + "님이 ";
            if (action.get(i).equals("Enter"))
                result += "들어왔습니다.";
            else
                result += "나갔습니다.";
            answer[i] = result;
        }
        
        return answer;
    }
}